---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# bacphene

<!-- badges: start -->
<!-- badges: end -->

The goal of bacphene is to generate a table of phenotype characteristics about your samples. Input is a table of Samples with counts per taxa. 

The package utilizes the official [BacDive R package] that accesses
bacterial phenotype data from [bacdive.org].

## Installation

You can install bacphene from the [github repo]. You also need to install the BacDive R package (if you want the full data - see below).

``` r
library(devtools)
install_github(repo = "scottdaniel/bacphene")
install.packages("BacDive", repos="http://R-Forge.R-project.org")
```

After this, you may register at bacdive.org [here](https://api.bacdive.dsmz.de/login).

### Recommended

Edit your $HOME/.Renviron file (you can open in R with `usethis::edit_r_environ()`) and add your bacdive credentials like so:
```
DSMZ_API_USER=your_email@something.com
DSMZ_API_PASSWORD=your_password
```

If you do not register, you will not be able to download the full information on strains in BacDive. Instead, what you will get is access to three data-frames: `bacdive_phenotypes`, `bacdive_susceptibility`, and `bacdive_enzymes`. Which look like this:

```{r}
library(bacphene)
head(bacdive_phenotypes)
```

```{r}
head(bacdive_susceptibility)
```

```{r}
head(bacdive_enzymes)
```

## Demo

Our test data-set is from Shen et al. 2021 and contains taxonomic counts from metagenomic shotgun alignments. See `?Shen2021` for a full description of the columns. With our test data and bacdive we can ask a few questions:

### Do the rectal swabs contain more oxygen tolerant organisms?

```{r}

library(tidyverse)

my_df <- Shen2021 %>%
  left_join(bacdive_phenotypes, by = "taxon") %>%
  mutate(prop = count / read_counts) %>%
  group_by(SampleID, aerobic_status) %>%
  filter(count > 0, !is.na(aerobic_status)) %>%
  summarise(n = n(), weighted_n = sum(prop) * n(), .groups = "drop") %>%
  ungroup()

sort_order <- my_df %>%
  group_by(SampleID) %>%
  mutate(prop = weighted_n / sum(weighted_n)) %>%
  filter(aerobic_status %in% "anaerobe") %>%
  arrange(prop) %>%
  pull(SampleID)

my_df %>%
  left_join(Shen2021 %>% select(SampleID, SampleType) %>% unique(), by = "SampleID") %>%
  mutate(SampleID = fct_relevel(SampleID, sort_order)) %>%
  mutate(aerobic_status = fct_relevel(aerobic_status, "anaerobe", "aerobe")) %>%
  
  ggplot(aes(x = SampleID, y = weighted_n, fill = aerobic_status)) +
  geom_bar(stat = "identity", position = "fill") +
  theme_bw() +
  theme(strip.text.x = element_text(size = 8),
        strip.background = element_blank(),
        axis.text.x = element_blank()) +
  facet_wrap(~SampleType, scales = "free") +
  scale_y_continuous(labels = scales::percent) +
  scale_fill_brewer(palette = "Set1") +
  labs(y = "Proportion of bacteria type weighted by abundance", x = "", fill = "Aerobic status", caption = "Each column represents an individual sample")

```

What are contributing most to these counts?

#### Top taxa in feces

```{r}

library(pander)

top_taxa_feces <- Shen2021 %>%
  left_join(bacdive_phenotypes, by = "taxon") %>%
  filter(SampleType %in% "Feces") %>%
  group_by(SampleType, aerobic_status, taxon) %>%
  summarise(total_count = sum(count), .groups = "drop") %>%
  mutate(percentage = paste(round((total_count / sum(total_count))*100), "%")) %>%
  slice_max(order_by = total_count, n = 10)

top_taxa_feces %>% pander(split.table = Inf, split.cells = 20, digits = 2)

```

#### Top taxa in rectal swab

```{r}

top_taxa_rectal <- Shen2021 %>%
  left_join(bacdive_phenotypes, by = "taxon") %>%
  filter(SampleType %in% "Rectal swab") %>%
  group_by(SampleType, aerobic_status, taxon) %>%
  summarise(total_count = sum(count), .groups = "drop") %>%
  mutate(percentage = paste(round((total_count / sum(total_count))*100), "%")) %>%
  slice_max(order_by = total_count, n = 10)

top_taxa_rectal %>% pander(split.table = Inf, split.cells = 20, digits = 2)

```

#### Is the aerobe / anaerobe ratio significantly different?

```{r}

totals_df <- my_df %>%
  pivot_wider(-n, names_from = "aerobic_status", values_from = "weighted_n") %>%
  mutate(log_aerobe_ratio = log(aerobe / anaerobe)) %>%
  left_join(Shen2021 %>% select(SampleID, SampleType) %>% unique(), by = "SampleID")

my_lm <- lm(log_aerobe_ratio ~ SampleType, data = totals_df)

summary(my_lm)

```

##### Graph

```{r}

totals_df %>%
  ggplot(aes(x = SampleType, y = log_aerobe_ratio)) +
  geom_boxplot()

```

Looks like there are slightly more aerobes compared to anaerobes in the rectal swabs but not enough to be statistically significant.

### Differences in gram-staining

```{r}



```


## References

When using BacDive for research please consider citing the following paper:

BacDive in 2019: bacterial phenotypic data for High-throughput biodiversity analysis.
Reimer, L. C., Vetcininova, A., Sardà Carbasse, J., Söhngen, C., Gleim, D., Ebeling, C., Overmann, J.
[Nucleic Acids Research; database issue 2019](https://academic.oup.com/nar/article/47/D1/D631/5106998).

Sample data set from:
Shen, T.-C. D. et al. (2021) ‘The Mucosally-Adherent Rectal Microbiota Contains Features Unique to Alcohol-Related Cirrhosis’, [Gut microbes, 13(1), p. 1987781.](https://www.ncbi.nlm.nih.gov/pubmed/34747331)

To cite this package enter `citation("bacphene")` in your R console.

[BacDive R package]: https://r-forge.r-project.org/R/?group_id=1573
[bacdive.org]: https://bacdive.dsmz.de
[github repo]: https://github.com/scottdaniel/bacphene
