---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# bacphene

<!-- badges: start -->
<!-- badges: end -->

The goal of bacphene is to generate a table of phenotype characteristics about your samples. Input is a table of Samples with counts per taxa. 

The package utilizes the official [BacDive R package] that accesses
bacterial phenotype data from [bacdive.org].

## Installation

You can install bacphene from the [github repo]. You also need to install the BacDive R package (if you want the full data - see below).

``` r
library(devtools)
install_github(repo = "scottdaniel/bacphene")
install.packages("BacDive", repos="http://R-Forge.R-project.org")
```

After this, you may register at bacdive.org [here](https://api.bacdive.dsmz.de/login).

### Recommended

Edit your $HOME/.Renviron file (you can open in R with `usethis::edit_r_environ()`) and add your bacdive credentials like so:
```
DSMZ_API_USER=your_email@something.com
DSMZ_API_PASSWORD=your_password
```

If you do not register, you will not be able to download the full information on strains in BacDive. Instead, what you will get is access to three data-frames: `bacdive_phenotypes`, `bacdive_susceptibility`, and `bacdive_enzymes`. Which look like this:

```{r}
library(bacphene)
head(bacdive_phenotypes)
```

```{r}
head(bacdive_susceptibility)
```

```{r}
head(bacdive_enzymes)
```

## Demo

Our test data-set is from Shen et al. 2021 and contains taxonomic counts from metagenomic shotgun alignments. See `?Shen2021` for a full description of the columns. With our test data and bacdive we can ask a few questions:

### Do the rectal swabs contain more oxygen tolerant organisms?

While this may seem like an obvious "yes", it still stands as a proof-of-concept.

```{r}

library(tidyverse)

my_df <- Shen2021 %>%
  left_join(bacdive_phenotypes, by = "taxon")

my_df %>%
  group_by(SampleType, aerobic_status) %>%
  filter(count > 0, !is.na(aerobic_status)) %>%
  summarise(n = n(), weighted_n = sum(count) * n(), .groups = "drop") %>%
  group_by(SampleType) %>%
  mutate(normalised_n = weighted_n / sum(weighted_n)) %>%
  ungroup() %>%
  ggplot(aes(x = SampleType, y = weighted_n, fill = aerobic_status)) +
    geom_bar(stat = "identity", position = "fill") +
  theme_bw() +
  theme(strip.text.x = element_text(size = 8),
        strip.background = element_blank(),
        axis.text.x = element_text(angle=45, hjust = 1)) +
  scale_y_continuous(labels = scales::percent) +
  labs(y = "Relative abundance", x = "", fill = "Aerobic status")

# Seems like there should still be more aerobic bacteria in the rectal swabs

```

```{r}

top_taxa <- my_df %>%
  group_by(SampleType, aerobic_status, taxon) %>%
  summarise(total_count = sum(count), .groups = "drop") %>%
  slice_max(order_by = total_count, n = 5)

```

## References

When using BacDive for research please consider citing the following paper:

BacDive in 2019: bacterial phenotypic data for High-throughput biodiversity analysis.
Reimer, L. C., Vetcininova, A., Sardà Carbasse, J., Söhngen, C., Gleim, D., Ebeling, C., Overmann, J.
[Nucleic Acids Research; database issue 2019](https://academic.oup.com/nar/article/47/D1/D631/5106998).

To cite this package enter `citation("bacphene")` in your R console.

[BacDive R package]: https://r-forge.r-project.org/R/?group_id=1573
[bacdive.org]: https://bacdive.dsmz.de
[github repo]: https://github.com/scottdaniel/bacphene
